spotpass/3ds-task-finder.js

```diff
@@ -0,0 +1,73 @@
+const https = require('node:https');
+const axios = require('axios');
+const fs = require('fs-extra');
+const { COUNTRIES, LANGUAGES } = require('./constants');
+const apps = require('./ctr-boss-apps.json');
+
+const NPFL_URL_BASE = 'https://npfl.c.app.nintendowifi.net/p01/filelist';
+const TASKS = ['FGONLYT', 'news', 'data', 'TASK00', '0000001']; // * For this example, I'm using some common task names that are present in many games, but there are other ways this can be used, such as for matching tasks across different regions for a single game.
+
+const httpsAgent = new https.Agent({
+	rejectUnauthorized: false,
+	cert: fs.readFileSync('./certs/wiiu-common.crt'),
+	key: fs.readFileSync('./certs/wiiu-common.key'),
+});
+
+async function main() {
+	for (const app of apps) {
+		for (const task of TASKS) {
+			// * Skip the task if the app already has it
+			let titleHasTask = false;
+			for (const appTask of app.tasks) {
+				if (task.toLowerCase() == appTask.toLowerCase()) {
+					titleHasTask = true;
+					break;
+				}
+			}
+
+			if (titleHasTask) {
+				continue;
+			}
+
+			// * Most BOSS apps are region-agnostic, but some require
+			// * specific combinations. Try every country/language
+			// * combination until a task is found
+			let taskFound = true;
+			for (const country of COUNTRIES) {
+				for (const language of LANGUAGES) {
+					if (await taskExists(app, task, country, language)) {
+						console.log(`Task ${task} found for app id ${app.app_id}`);
+						app.tasks.push(task);
+						await fs.writeJSONSync('./ctr-boss-apps.json', apps, {
+							spaces: '\t'
+						});
+
+						taskFound = true;
+						break;
+					}
+				}
+			}
```

Comment:
`break` in this instance will only break out of the inner loop, not the outer loop. So even when a task is found it will continue looping through any remaining countries, despite not needing to do so.

There is a number of ways in JavaScript to fix this

### 1. Labels
Simply wrapping the outer loop in a [label](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/label) and using `break` to jump out of it will exit both loops:

```suggestion
			check_locales: for (const country of COUNTRIES) {
				for (const language of LANGUAGES) {
					if (await taskExists(app, task, country, language)) {
						console.log(`Task ${task} found for app id ${app.app_id}`);
						app.tasks.push(task);
						await fs.writeJSONSync('./ctr-boss-apps.json', apps, {
							spaces: '\t'
						});

						taskFound = true;
						break check_locales;
					}
				}
			}
```

### 2. Using a standard `for` loop with an extra condition
`for` loops in JavaScript can do more than one thing in all 3 sections. In this case, creating an extra variable in the `initialization` section and doing an extra check in the `condition` section:

```suggestion
			for (let i = 0, deepExit = false; i < COUNTRIES.length && !deepExit; i++) {
				const country = COUNTRIES[i];

				for (const language of LANGUAGES) {
					if (await taskExists(app, task, country, language)) {
						console.log(`Task ${task} found for app id ${app.app_id}`);
						app.tasks.push(task);
						await fs.writeJSONSync('./ctr-boss-apps.json', apps, {
							spaces: '\t'
						});

						taskFound = true;
						deepExit = true;
						break;
					}
				}
			}
```

Or reusing the `taskFound` variable to do the same job.

### 3. Wrapping the loop in a function
Wrapping the loop in a function, either self-executing or a standard function declaration, and then using `return` to exit:

```suggestion
			// Self-executing
			(() => {
				for (const country of COUNTRIES) {
					for (const language of LANGUAGES) {
						if (await taskExists(app, task, country, language)) {
							console.log(`Task ${task} found for app id ${app.app_id}`);
							app.tasks.push(task);
							await fs.writeJSONSync('./ctr-boss-apps.json', apps, {
								spaces: '\t'
							});
	
							taskFound = true;
							return;
						}
					}
				}
			})();
```

```suggestion
			// Standard function declaration
			function checkTaskLocales() {
				for (const country of COUNTRIES) {
					for (const language of LANGUAGES) {
						if (await taskExists(app, task, country, language)) {
							console.log(`Task ${task} found for app id ${app.app_id}`);
							app.tasks.push(task);
							await fs.writeJSONSync('./ctr-boss-apps.json', apps, {
								spaces: '\t'
							});
	
							return true;
						}
					}
				}
				
				return false;
			}
```

Reply:
That's weird because I thought I fixed that. I'll have a look.